### PR TITLE
Feature/issue 13/add usecase newslist

### PR DESCRIPTION
**Pull request by timofeylyadinsky**
domain for news list screen
1. add use case to get news articles from api

### CODE REVIEW

File: app/src/main/java/com/example/newsapp/domain/NewsListUseCases.kt
Comment (MaxSkr-iTechArt) [inline app/src/main/java/com/example/newsapp/domain/NewsListUseCases.kt:1]: Refer to this article to implement a better solution for handling network responses
https://proandroiddev.com/modeling-retrofit-responses-with-sealed-classes-and-coroutines-9d6302077dfe

**Comment by timofeylyadinsky** [comment]
1. add mapping to object from data to domain
2. add sealed class network result to check status of response
3. add flow to fetching data

File: app/src/main/java/com/example/newsapp/data/repository/NewsRepository.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import com.example.newsapp.data.api.ApiService
+import com.example.newsapp.data.api.NetworkResult
+import com.example.newsapp.data.entity.NewsDto
+import retrofit2.Response
+import javax.inject.Inject
+
+class NewsRepository @Inject constructor(
+    private val apiService: ApiService
+) {
+    suspend fun<T: Any> handleApi(
```
Comment (MaxSkr-iTechArt) [inline app/src/main/java/com/example/newsapp/data/repository/NewsRepository.kt]: I think you can make it a util function being available for all repositories.

File: app/src/main/java/com/example/newsapp/data/repository/NewsRepository.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+                NetworkResult.Success(body)
+            } else {
+                NetworkResult.Error(response.code(), response.message())
+            }
+        } catch (e: Exception) {
+            NetworkResult.Error(e.hashCode(), e.message)
+        }
+    }
+
+    suspend fun dataSource() :NetworkResult<NewsDto> = handleApi { apiService.getNews() }
```
Comment (MaxSkr-iTechArt) [inline app/src/main/java/com/example/newsapp/data/repository/NewsRepository.kt]: DOn't forget to use code formatting :)

File: app/src/main/java/com/example/newsapp/domain/NewsListUseCases.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class GetNewsResponseUseCase @Inject constructor(
+    private val newsRepository: NewsRepository,
+    @IoDispatcher private val ioDispatcher: CoroutineDispatcher
+) {
+    suspend operator fun invoke(): NetworkResult<NewsDto> = withContext(ioDispatcher) {
```
Comment (MaxSkr-iTechArt) [inline app/src/main/java/com/example/newsapp/domain/NewsListUseCases.kt]: Let's discuss

File: app/src/main/java/com/example/newsapp/domain/entity/NewsData.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.example.newsapp.domain.entity
+
+data class NewsData(
+    val news: List<Article> = listOf(),
+    val errorMessage: String? = null
```
Comment (MaxSkr-iTechArt) [inline app/src/main/java/com/example/newsapp/domain/entity/NewsData.kt:5]: Should we introduce the loading status here, i.e. isLoading?

**Review by MaxSkr-iTechArt — APPROVED** [review]

**Pull request merged**